powershell.html.markdown

```diff
@@ -133,89 +373,54 @@ switch($val) {
+    Write-Output $Error.ExceptionMessage
+}
+finally {
+    Write-Output "We can clean up resources here"
 }
 
 
-### Providers
-# List files and directories in the current directory
-ls # or `dir`
-cd ~ # goto home
-
-Get-Alias ls # -> Get-ChildItem
-# Uh!? These cmdlets have generic names because unlike other scripting
-# languages, PowerShell does not only operate in the current directory.
-cd HKCU: # go to the HKEY_CURRENT_USER registry hive
-
-# Get all providers in your session
-Get-PSProvider
-
-
-### Pipeline
-# Cmdlets have parameters that control their execution:
-Get-ChildItem -Filter *.txt -Name # Get just the name of all txt files
-# Only need to type as much of a parameter name until it is no longer ambiguous
-ls -fi *.txt -n # -f is not possible because -Force also exists
-# Use `Get-Help Get-ChildItem -Full` for a complete overview
+# Writing to a file
+$contents = @{"aa"= 12 
+             "bb"= 21}
+$contents | Export-CSV "$env:HOMEDRIVE\file.csv" # writes to a file
 
-# Results of the previous cmdlet can be passed to the next as input.
-# `$_` is the current object in the pipeline object.
-ls | Where-Object { $_.Name -match 'c' } | Export-CSV export.txt
-ls | ? { $_.Name -match 'c' } | ConvertTo-HTML | Out-File export.html
+$contents = "test string here"
+$contents | Out-File "$env:HOMEDRIVE\file.txt" # writes to another file
 
-# If you get confused in the pipeline use `Get-Member` for an overview
-# of the available methods and properties of the pipelined objects:
-ls | Get-Member
-Get-Date | gm
+# Read file contents and convert to json
+Get-Content "$env:HOMEDRIVE\file.csv" | ConvertTo-Json
 
-# ` is the line continuation character. Or end the line with a |
-Get-Process | Sort-Object ID -Descending | Select-Object -First 10 Name,ID,VM `
-	| Stop-Process -WhatIf
 
-Get-EventLog Application -After (Get-Date).AddHours(-2) | Format-List
 
-# Use % as a shorthand for ForEach-Object
-('a','b','c') | ForEach-Object `
-	-Begin { "Starting"; $counter = 0 } `
-	-Process { "Processing $_"; $counter++ } `
-	-End { "Finishing: $counter" }
+####################################################
+## 4. Functions
+####################################################
 
-# Get-Process as a table with three columns
-# The third column is the value of the VM property in MB and 2 decimal places
-# Computed columns can be written more verbose as:
-# `@{name='lbl';expression={$_}`
-ps | Format-Table ID,Name,@{n='VM(MB)';e={'{0:n2}' -f ($_.VM / 1MB)}} -autoSize
+# Use "function" to create new functions
+# Keep the Verb-Noun naming convention for functions
+function Add-Numbers {
+ $args[0] + $args[1]
+}
 
+Add-Numbers 1 2 # => 3
 
-### Functions
-# The [string] attribute is optional.
-# Function names should follow Verb-Noun convention
-function Get-Foo([string]$name) {
-	echo "Hey $name, have a function"
+# Calling functions with parameters
+function Add-ParamNumbers {
+ param( [int]$FirstNumber, [int]$SecondNumber )
+ $FirstNumber + $SecondNumber
```

Comment:
How about using `Add-ParamNumbers([int]$FirstNumber, [int]$SecondNumber)`, which would make it more similar to other language.

Reply:
There are later examples that illustrate this. I think in production you'd just go with whatever your team does.